### PR TITLE
[v1.x] License updates 

### PR DESCRIPTION
**Pull request by josephevans**
## Description ##
This PR fixes license issues identified in https://github.com/apache/incubator-mxnet/issues/20616 and more:

- Update Caffe license description in LICENSE to BSD-2 Clause with Caffe Copyright Notice and Disclaimer
- Fix category for 3rdparty/mkldnn/src/common/primitive_hashing.hpp (is Apache licensed, not Boost 1.0)
- Sort files in LICENSE in each section alphabetically
- Remove duplicate entries in LICENSE
- Update third-party software in mkldnn in LICENSE, per 3rdparty/mkldnn/THIRD-PARTY-PROGRAMS
- Remove incorrectly-added ASF headers on MIT-licensed files
- Add license text for each subcomponent that will be distributed in binary form to licenses/LICENSE.*.txt
  - This ensures we comply with licensing requirements for binary distributions

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -324,14 +327,6 @@
...
-    =======================================================================================
-    Apache-2.0 license + 3-clause BSD license
-    =======================================================================================
-
-    python/mxnet/onnx/mx2onnx/_export_onnx.py
-    python/mxnet/onnx/mx2onnx/_op_translations/_op_translations_opset12.py
-    python/mxnet/onnx/mx2onnx/_op_translations/_op_translations_opset13.py
-
     =======================================================================================
     Apache-2.0 license + MIT License
```
[the remark this answers is not in the document]
Comment (josephevans) [reply LICENSE]: OK, in order to make the licensing clear for this particular issue, I've moved the single function (LayerNormCPUKernel) into it's own file (`layer_norm_cpu.h`). This file is MIT-licensed, while the remaining code will still be under the ASF-2.0 license.

**Comment by josephevans** [comment]
> Under no circumstances do we want multiple licenses regardless of where they are in LICENSE. Let's use it to detect whether there are other issues with dual license.

Got it. I updated the tool to error out when it finds a file with multiple licenses.

File: licenses/LICENSE.blockingconcurrentqueue.txt [hunk 2]
```diff
@@ -0,0 +1,26 @@
+©2015-2016 Cameron Desrochers. Distributed under the terms of the simplified
+BSD license, available at the top of concurrentqueue.h.
+
+Uses Jeff Preshing's semaphore implementation (under the terms of its
+separate zlib license, embedded below).
+
+
+zlib license
```
Comment (szha) [inline licenses/LICENSE.blockingconcurrentqueue.txt:8]: how do we avoid including this in source distribution?

Comment (josephevans) [reply licenses/LICENSE.blockingconcurrentqueue.txt:8]: This is part of TVM. This is the same way they include the license for this file.

Comment (josephevans) [reply licenses/LICENSE.blockingconcurrentqueue.txt:8]: See https://github.com/apache/tvm/blob/main/LICENSE#L223-L226 and https://github.com/apache/tvm/blob/main/licenses/LICENSE.blockingconcurrentqueue.txt for references in TVM.

**Review by szha — APPROVED** [review]
@josephevans thanks for addressing the issues!

**Review by mseth10 — APPROVED** [review]
LGTM

**Pull request merged**